index.js

```diff
@@ -20,39 +20,36 @@
 //. purpose: the [`typeof`][1] operator and [`Object.prototype.toString`][2].
 //. Each has pros and cons, but neither supports user-defined types.
 //.
-//. This package specifies an [algorithm][3] for deriving a _type identifier_
-//. from any JavaScript value, and exports an implementation of the algorithm.
-//. Authors of algebraic data types may follow this specification in order to
-//. make their data types compatible with the algorithm.
+//. This package exports a function which derives a _type identifier_ from any
+//. JavaScript value, and a [specification][3] for customizing the type
+//. identifier of a value.
 //.
-//. ### Algorithm
+//. ### Specification
 //.
-//. 1.  Take any JavaScript value `x`.
+//. For an algebraic data type to be compatible with the algorithm:
 //.
-//. 2.  If `x` is `null` or `undefined`, go to step 6.
-//.
-//. 3.  If `x.constructor` evaluates to `null` or `undefined`, go to step 6.
-//.
-//. 4.  If `x.constructor.prototype === x`, go to step 6. This check prevents a
-//.     prototype object from being considered a member of its associated type.
+//.   - every member of the type MUST have a `constructor` property pointing
+//.     to an object known as the _type representative_;
 //.
-//. 5.  If `typeof x.constructor['@@type']` evaluates to `'string'`, return
-//.     the value of `x.constructor['@@type']`.
+//.   - the type representative MUST have a `@@type` property;
 //.
-//. 6.  Return the [`Object.prototype.toString`][2] representation of `x`
-//.     without the leading `'[object '` and trailing `']'`.
+//.   - the type representative's `@@type` property (the _type identifier_)
+//.     MUST be a string primitive and SHOULD have format:
+//.     `'<namespace>:<name>[@<version>]'`, where:
 //.
-//. ### Compatibility
+//.       - The `namespace` MUST be a [valid npm package name][4] which SHOULD
+//.         equal the npm package name which defines the type; and
```

Comment:
Do we keep the rule that states it must be a valid npm package name? I've removed the validations from the RegExp, and I think there are two ways to go:

1. We depend on [`validate-npm-package-name`](https://github.com/npm/validate-npm-package-name) to validate our "namespace". If the validation fails, we concatenate the namespace part back to the name part in our result.
1. We remove this rule from the specification (keeping the SHOULD-part) and allow any `non-slash` string to be the namespace.